### PR TITLE
Alan Shepard, AKA our noble steward

### PR DESCRIPTION
**Pull request by arigoldx**
Now Allie (we're pretty close so I call 'im Allie) can monitor files in repos for you:

    alan monitor {repo} for {path} - path is a regex (the script prepends a "/")
    alan stop all monitoring period - for everyone, so know what you're doing
    alan stop monitoring {path} in {repo}
    alan show steward - for everyone

### CODE REVIEW

File: scripts/pull-request-monitor.coffee [hunk 1]
```diff
@@ -0,0 +1,109 @@
+# Description:
+#   This monitors if a Pull Request is opened, reopened or synchronized and
+#   notified the user if any files referenced by the PR match the given path.
+#
+# Dependencies:
```
Comment (riveramj) [inline scripts/pull-request-monitor.coffee]: you require `util` below. 
https://github.com/github/hubot-scripts#documentation

File: scripts/pull-request-monitor.coffee [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+#   QQQ hmm...
+#
+# Configuration:
+#   QQQ hmm...
```
Comment (riveramj) [inline scripts/pull-request-monitor.coffee]: you require the github token below
https://github.com/github/hubot-scripts#documentation

File: scripts/pull-request-monitor.coffee [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+#
+# Commands:
+#   hubot monitor {repo} for {path} - path is a regex (the script prepends a "/")
+#   hubot stop all monitoring period - for everyone, so know what you're doing
+#   hubot stop monitoring {path} in {repo}
+#   hubot show steward - for everyone
+# 
+# Author: 
+#   hacklanta
```
Comment (riveramj) [inline scripts/pull-request-monitor.coffee]: so prob not hacklanta?

File: scripts/pull-request-monitor.coffee [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+
+Util = require 'util'
+
+GITHUB_TOKEN = process.env['GITHUB_TOKEN']
+
+module.exports = (robot) ->
+
+#
+#
```
Comment (riveramj) [inline scripts/pull-request-monitor.coffee]: can remove these lines

File: scripts/pull-request-monitor.coffee [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  robot.respond /monitor (\S+) for (\S+$)/i, (msg) ->
+    getAffairInOrder(msg)
+
+  getAffairInOrder = (msg) ->
```
Comment (riveramj) [inline scripts/pull-request-monitor.coffee]: `addToAffairs` maybe?

`getAffairInOrder` seems like some sort of cleanup job

File: scripts/pull-request-monitor.coffee [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    steward[repo] ||= []
+
+    steward[repo].push { path: path, user: msg.message.user }
+      
+    msg.send "Okay. I'm monitoring #{path} in #{repo}."
+
+    robot.brain.set 'steward', steward
+
+#
+#
```
Comment (riveramj) [inline scripts/pull-request-monitor.coffee]: ...

File: scripts/pull-request-monitor.coffee [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  robot.respond /stop monitoring (\S+) in (\S+$)/i, (msg) ->
+    stopSomeMonitoring(msg)
```
Comment (riveramj) [inline scripts/pull-request-monitor.coffee]: this seems like `getAffairInOrder` maybe

File: scripts/pull-request-monitor.coffee [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    
+    steward[repo] = steward[repo].filter (monitoredPath) ->
+      (monitoredPath.path != path) && (monitoredPath.user != user)
+
+    robot.brain.set 'steward', steward
+
+    msg.send "Okay. I'm no longer montioring for #{path} in #{repo} for you."
+
+#
+#
```
Comment (riveramj) [inline scripts/pull-request-monitor.coffee]: am i missing something here? pls tell me if i am :-P

File: scripts/pull-request-monitor.coffee [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+  robot.respond /stop all monitoring period/i, (msg) ->
+    robot.brain.set 'steward', {}
+
+    msg.send "Monitor what? There's nothing to monitor. ;-)"
```
Comment (riveramj) [inline scripts/pull-request-monitor.coffee]: i like it

File: scripts/pull-request-monitor.coffee [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+#
+#
+  robot.respond /show steward/i, (msg) ->
```
Comment (riveramj) [inline scripts/pull-request-monitor.coffee]: i'm ~ on this command. like we can use it. not saying that. just not sure the wording is right.

File: scripts/pull-request-monitor.coffee [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    steward = robot.brain.get('steward') || {}
+    msg.send "steward: #{JSON.stringify(steward)}"
+
+#
+#
+  robot.router.post '/pull-request-activity', (req, res) ->
```
Comment (riveramj) [inline scripts/pull-request-monitor.coffee]: maybe `steward/pull-request` to allow us to monitor other things?

File: scripts/pull-request-monitor.coffee [hunk 12]
```diff
@@ -0,0 +1,109 @@
+# Description:
+#   This monitors if a Pull Request is opened, reopened or synchronized and
```
Comment (riveramj) [inline scripts/pull-request-monitor.coffee]: not `synchronized`

File: scripts/pull-request-monitor.coffee (hunk 3 above)
Comment (arigoldx) [inline scripts/pull-request-monitor.coffee]: right. but who? shall I just put both of our names?

File: scripts/pull-request-monitor.coffee (hunk 4 above)
Comment (arigoldx) [inline scripts/pull-request-monitor.coffee]: ah, I kinda liked 'em & hoped the'd squeak by. they separate the different logical components. I'll kill 'em.

File: scripts/pull-request-monitor.coffee [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+      number = req.body.pull_request.number
+      action = req.body.action
+      repo = req.body.repository.name
+      
+      steward = robot.brain.get('steward')
+
+      if steward && steward[repo]
+        if action in ["opened", "reopened", "closed"]
+          robot
+            .http("https://api.github.com/repos/elemica/" + repo + "/pulls/" + number + "/files")
```
Comment (riveramj) [inline scripts/pull-request-monitor.coffee]: if we are going to open source this, we need to remove elemica.
might be worth ppl doing `company/repo` name now. or making it a one time set thing.
```
if steward.owner is set 
  steward.owner + repo
else 
  repo (which would have to be owner/repo)
```

File: scripts/pull-request-monitor.coffee (hunk 3 above)
Comment (riveramj) [inline scripts/pull-request-monitor.coffee]: sure. can use github names if you want.

Comment (riveramj) [inline scripts/pull-request-monitor.coffee]: ^

File: scripts/pull-request-monitor.coffee (hunk 4 above)
Comment (riveramj) [inline scripts/pull-request-monitor.coffee]: eh we can leave them then.

File: scripts/pull-request-monitor.coffee [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+            .header('authorization', "token #{GITHUB_TOKEN}")
+            .get() (err, res, body) ->
+              if err
+                robot.send "Encountered an error monitoring PR#{number} :( ${err}"
+              else
+                files = JSON.parse(body)
+                length = files.length
```
Comment (riveramj) [inline scripts/pull-request-monitor.coffee]: you dont use this line

File: scripts/pull-request-monitor.coffee [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+                for file in files
```
Comment (riveramj) [inline scripts/pull-request-monitor.coffee]: where is a scala for comprehension when you need one.
may be a better way to do `for -> for -> if`

File: scripts/pull-request-monitor.coffee [hunk 16]
```diff
@@ -0,0 +1,98 @@
+# Description:
```
Comment (riveramj) [inline scripts/pull-request-monitor.coffee]: maybe the file name should be `pull-request-steward`? or `github-steward`? or just `steward.coffee`

File: scripts/pull-request-monitor.coffee (hunk 3 above)
Comment (arigoldx) [inline scripts/pull-request-monitor.coffee]: Author:
      riveramj arigoldx
?

Comment (riveramj) [inline scripts/pull-request-monitor.coffee]: two lines

File: scripts/pull-request-monitor.coffee (hunk 13 above)
Comment (arigoldx) [inline scripts/pull-request-monitor.coffee]: yes good point.

we could do a
    alan set steward owner

and have an error telling the user that the owner isn't set and to pretty please set it?

Comment (arigoldx) [inline scripts/pull-request-monitor.coffee]: I'm gonna do that ^

Comment (Shadowfiend) [inline scripts/pull-request-monitor.coffee]: Pulling from an env variable is also a reasonable solution.

File: scripts/pull-request-monitor.coffee (hunk 15 above)
Comment (arigoldx) [inline scripts/pull-request-monitor.coffee]: seriously.. I had trouble shaking the desire to include underscore to get the functional stuff in 'er.

I'm all :ear: but I'll leave it like this for now...

File: scripts/pull-request-monitor.coffee (hunk 13 above)
Comment (arigoldx) [inline scripts/pull-request-monitor.coffee]: I like the env variable better but I dunno how I can test that..seems like I'd need access to the box Alan's on as well as probably update some script that initialized env variables on build. @Shadowfiend?

File: scripts/pull-request-monitor.coffee (hunk 10 above)
Comment (arigoldx) [inline scripts/pull-request-monitor.coffee]: maybe `show steward affairs`?

Comment (arigoldx) [inline scripts/pull-request-monitor.coffee]: or just `show affairs`?

File: scripts/pull-request-monitor.coffee (hunk 13 above)
Comment (riveramj) [inline scripts/pull-request-monitor.coffee]: i have a script that i'll add the env to. for local, you just have to `export ENV=value` before you run hubot, same as the github one.

File: scripts/pull-request-monitor.coffee (hunk 10 above)
Comment (riveramj) [inline scripts/pull-request-monitor.coffee]: `show affairs` words for me.

File: scripts/steward.coffee [hunk 17]
```diff
@@ -0,0 +1,96 @@
+# Description:
+#   This monitors if a Pull Request is opened, reopened or closed and
+#   notified the user if any files referenced by the PR match the given path.
+#
+# Dependencies:
+#   none
+#
+# Configuration:
+#   GITHUB_TOKEN - sent in Http GET header
```
Comment (riveramj) [inline scripts/steward.coffee]: need to add steward organization here

Comment (arigoldx) [inline scripts/steward.coffee]: STEWARD_ORGANIZATION - your GitHub organization or username
?

Comment (riveramj) [inline scripts/steward.coffee]: sounds good.

**Comment by riveramj** [comment]
one thing that would be nice is a link to the PR and/or the file in the PR.

that can be added after the fact. I'll open an issue/

File: scripts/steward.coffee [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+    steward = robot.brain.get('steward') || {}
+    steward[repo] ||= []
+
+    steward[repo].push { path: path, user: msg.message.user }
+      
+    msg.send "Okay. I'm monitoring #{path} in #{repo}."
+
+    robot.brain.set 'steward', steward
+
+  robot.respond /stop monitoring (\S+) in (\S+$)/i, (msg) ->
```
Comment (riveramj) [inline scripts/steward.coffee]: lets change this to `stop monitoring {repo} for {path}` to keep it uniform with the monitoring command.

Comment (riveramj) [inline scripts/steward.coffee]: i would also make it `monitor|monitoring`

**Comment by riveramj** [comment]
something that I caused :-(

if i say monitor `configuration`, it wont match `partner-configuration`.

**Comment by riveramj** [comment]
also this continues to slightly bug me. If you add a file after a PR was opened that matches your filter, you won't know until the PR is closed.

**Comment by arigoldx** [comment]
so given that `configuration` vs `partner-configuration` of yours are you suggesting we remove the leading "/"?

**Comment by riveramj** [comment]
i guess. and have them add the `/` if they want it to be an exact match for folders.

**Comment by arigoldx** [comment]
yeah I hear you about wanting notification for already opened PRs but:

1. it's just those open PRs. from then on, until the sun explodes, you'll get notifications
2. the alternative is worse - to get notifications anytime something syncs to that PR
3. we can always add a flag to check for all synchronizations:

    Alan monitor mercury for CloudsearchFacadeAPI on sync

**Comment by riveramj** [comment]
> it's just those open PRs. from then on, until the sun explodes, you'll get notifications

what if I add commits after a PR is opened. Like how I see it:

> (Mike opens a Scribe PR w/o documentation)
> Benno: Mike where is the updated doc for the new verb?
> Mike: Oh crap. 1 Sec. Pushed
> Benno: Excellent.
> Benno: Seems to be no complaints on that new documentation. Merged.
> Dani: WTF just happened and why didnt I know bout this before merge?

**Comment by riveramj** [comment]
dunno if that is really an issue. just my thoughts. we can skip it for now i'm sure. it notifies you on close, which works all the same until someone actually complains.

**Comment by arigoldx** [comment]
well if it's to be used for real-time notification of files then we prolly need to just include synchronization and let the user remove the notification if bugs 'em too much (or quiet it with `stop synchronizing for ...`.)

I was under the impression that it was for digest-style notifications...

**Comment by riveramj** [comment]
dunno. user feedback will tell us.

**Comment by riveramj** [comment]
aight lets go!

**Pull request merged**